lib/asyncSender.ts

```diff
@@ -0,0 +1,181 @@
+   * Default: `20 seconds`.
+   */
+  messageTimeoutInSeconds: number;
+  /**
+   * @property {Map<number, PromiseLike} deliveryDispositionMap Maintains a map of delivery of
+   * messages that are being sent. It acts as a store for correlating the dispositions received
+   * for sent messages.
+   */
+  deliveryDispositionMap: Map<number, PromiseLike> = new Map<number, PromiseLike>();
+
+  constructor(session: Session, sender: RheaSender, options: AsyncSenderOptions = {}) {
+    super(session, sender, options);
+    this.messageTimeoutInSeconds = options.messageTimeoutInSeconds || 20;
+
+    const onSuccess = (context: EventContext) => {
+      const id = context.delivery!.id;
+      if (this.deliveryDispositionMap.has(context.delivery!.id)) {
+        const promise = this.deliveryDispositionMap.get(id) as PromiseLike;
+        clearTimeout(promise.timer);
+        log.sender(
+          "[%s] Event: 'Accepted', Found the delivery with id %d in the map of " +
+          "sender '%s' on amqp session '%s' and cleared the timer.",
+          this.connection.id, id, this.name, this.session.id
+        );
+        const deleteResult = this.deliveryDispositionMap.delete(id);
+        log.sender(
+          "[%s] Event: 'Accepted', Successfully deleted the delivery with id %d from " +
+          "the map of sender '%s' on amqp session '%s' and cleared the timer: %s.",
+          this.connection.id, id, this.name, this.session.id, deleteResult
+        );
+        return promise.resolve(context.delivery);
+      }
+    };
+
+    const onFailure = (eventName: string, id: number, error?: Error) => {
+      if (this.deliveryDispositionMap.has(id)) {
+        log.sender(
+          "[%s] Event: '%s', Found the delivery with id %d in the map of " +
+          "sender '%s' on amqp session '%s' and cleared the timer.",
+          this.connection.id, eventName, id, this.name, this.session.id
+        );
+        const promise = this.deliveryDispositionMap.get(id) as PromiseLike;
+        clearTimeout(promise.timer);
+        const deleteResult = this.deliveryDispositionMap.delete(id);
+        log.sender(
+          "[%s] Event: '%s', Successfully deleted the delivery with id %d from the " +
+          " map of sender '%s' on amqp session '%s' and cleared the timer: %s.",
+          this.connection.id, eventName, id, this.name, this.session.id, deleteResult
+        );
+        const msg = `Sender '${this.name}' on amqp session '${this.session.id}', received a ` +
+          `'${eventName}' disposition. Hence we are rejecting the promise.`;
+        const err = error || new Error(msg);
+        log.sender("[%s] %s", this.connection.id, msg);
+        return promise.reject(err);
+      }
+    };
+
+    const onError = (eventName: string, error?: Error) => {
+      for (const id of this.deliveryDispositionMap.keys()) {
+        onFailure(eventName, id, error);
+      }
+    };
+
+    this.on(SenderEvents.accepted, onSuccess);
```

Comment:
To be consistent with other event handlers which take in just what is needed rather than the entire context, can we pass the `context.delivery` and  to `onSuccess` here? Makes code reading easier

Reply:
Done.